src/JsonSchema/Constraints/Constraint.php

```diff
@@ -22,48 +22,31 @@
  */
 abstract class Constraint implements ConstraintInterface
 {
-    protected $schemaStorage;
-    protected $checkMode = self::CHECK_MODE_NORMAL;
-    protected $uriRetriever;
     protected $errors = array();
     protected $inlineSchemaProperty = '$schema';
 
     const CHECK_MODE_NORMAL = 1;
     const CHECK_MODE_TYPE_CAST = 2;
 
     /**
-     * @var null|Factory
+     * @var Factory
      */
     private $factory;
 
     /**
-     * @param int $checkMode
-     * @param SchemaStorage $schemaStorage
-     * @param UriRetrieverInterface $uriRetriever
      * @param Factory $factory
      */
-    public function __construct(
-        $checkMode = self::CHECK_MODE_NORMAL,
-        SchemaStorage $schemaStorage = null,
-        UriRetrieverInterface $uriRetriever = null,
-        Factory $factory = null
-    ) {
-        $this->checkMode     = $checkMode;
-        $this->uriRetriever  = $uriRetriever;
-        $this->factory       = $factory;
-        $this->schemaStorage = $schemaStorage;
+    public function __construct(Factory $factory = null)
```

Comment:
you should consider providing a BC layer for the constructor signature. Otherwise, this force to bump to 4.0 already. Supporting the old signature with a deprecation warning (building a factory from the older arguments) would make the upgrade path easier